crates/polars-parquet/src/parquet/encoding/bitpacked/pack.rs

```diff
@@ -1,81 +1,87 @@
 /// Macro that generates a packing function taking the number of bits as a const generic
 macro_rules! pack_impl {
-    ($t:ty, $bytes:literal, $bits:tt) => {
-        pub fn pack<const NUM_BITS: usize>(input: &[$t; $bits], output: &mut [u8]) {
+    ($t:ty, $bytes:literal, $bits:tt, $bits_minus_one:tt) => {
+        pub unsafe fn pack<const NUM_BITS: usize>(input: &[$t; $bits], output: &mut [u8]) {
             if NUM_BITS == 0 {
                 for out in output {
                     *out = 0;
                 }
                 return;
             }
-            assert!(NUM_BITS <= $bytes * 8);
+            assert!(NUM_BITS <= $bits);
             assert!(output.len() >= NUM_BITS * $bytes);
 
-            let mask = match NUM_BITS {
-                $bits => <$t>::MAX,
-                _ => ((1 << NUM_BITS) - 1),
-            };
+            let input_ptr = input.as_ptr();
+            let mut output_ptr = output.as_mut_ptr() as *mut $t;
+            let mut out_register: $t = read_unaligned(input_ptr);
 
-            for i in 0..$bits {
-                let start_bit = i * NUM_BITS;
-                let end_bit = start_bit + NUM_BITS;
-
-                let start_bit_offset = start_bit % $bits;
-                let end_bit_offset = end_bit % $bits;
-                let start_byte = start_bit / $bits;
-                let end_byte = end_bit / $bits;
-                if start_byte != end_byte && end_bit_offset != 0 {
-                    let a = input[i] << start_bit_offset;
-                    let val_a = <$t>::to_le_bytes(a);
-                    for i in 0..$bytes {
-                        output[start_byte * $bytes + i] |= val_a[i]
-                    }
+            seq_macro::seq!(i in 1..$bits_minus_one {
```

Comment:
I can imagine this compiles to a lot of code. How does it perform if we leave this a loop?

Reply:
That's a good point. In my microbenchmark, I got around 50 ns/iter with a loop, more than 10x slower than when unrolled. I honestly did not expect it to account for so much of the improvement. In a release build, it adds ~100 KB to the binary size and has a negligible effect on build time. Do you think that is an acceptable trade-off?

Reply:
Right.. Let's leave it for now. Can you add a comment about the microbenchmark you ran above this macro? It is good information for future refactors.